### ISSUE TITLE
defaults() incorrectly processes empty collections

### ISSUE THREAD

**Issue by vladimirtiukhtin** [hashicorp/terraform#28372]
### Terraform Version
```
Terraform v0.15.0
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.36.0
```

### Steps to Reproduce
Declare a variable with a type "object" and make one of its attributes optional
```
variable "node_groups" {
  type = list(object(
    {
      ...
      additional_security_groups  = optional(list(string))
      ...
    }
  ))
```
Then in a resource try to reference that using "defaults()" function, but as a default value specify an empty list (for example because respective resource does not accept "null" only empty or none-empty lists)

```
resource "aws_launch_template" "node_group" {
  for_each = {
    for node_group in var.node_groups : node_group["name"] => node_group
  }
  ...
  network_interfaces {
    security_groups = compact(
      concat([aws_security_group.eks_worker_nodes_default_sg.id], defaults(each.value["additional_security_groups"], []))
    )
    associate_public_ip_address = each.value["associate_public_ip_address"]
  }
  ...
}
```
Terraform apply results in an error

```
Invalid value for "defaults" parameter: invalid default value for string: string required.
```
Replacing empty list with something meaningful solves the problem. So the bug is in specifying exactly empty collections. Ps I did not check how it behaves on maps and other types

**Comment by alisdair** [comment]
Thanks for reporting! This looks like a duplicate of #28352. [Please see my comment on that issue explaining what's happening here](https://github.com/hashicorp/terraform/issues/28352#issuecomment-819541140).

If I've misunderstood the report, please let me know.

**Comment by vladimirtiukhtin** [comment]
Hi @alisdair. Thanks for the prompt response. Unfortunately I cannot find a way around. I tried suggested approach, but it seems like if a var gets populated outside the module, it breaks plan logic. For example:
```
locals {
  node_groups = defaults(
    var.node_groups,
    {
      subnet_ids = var.subnet_ids
    }
  )
}
```

Ends up with 
```
var.subnet_ids is a list of string, known only after apply
```

**Comment by alisdair** [comment]
It looks like you are trying to assign a list value as a default for a list attribute, which is the cause of this error. While I understand that is an intuitive thing to want to do, this is not how the `defaults` function was designed to work. [From the documentation](https://www.terraform.io/docs/language/functions/defaults.html):

> *Collection types* (`list`, `map`, and `set` types): Terraform will visit each of the collection elements in turn and apply defaults to them. In this case the default value is only a single value to be applied to all elements of the collection, so it must have a type compatible with the collection's element type rather than with the collection type itself.

That means that there is no way (that I can think of) to use `defaults` to express something like "for the `node_groups` list attribute, if it is unset, use this other list of strings instead". Depending on exactly how you use this local value, [the `coalescelist` function](https://www.terraform.io/docs/language/functions/coalescelist.html) might be applicable here.

While I don't think this is a bug in Terraform, the `defaults` function and the optional object attributes syntax are still experimental, so we're working on the design. [If you could share your use case in some detail on the discussion forum](https://discuss.hashicorp.com/c/terraform-core/27), that would help us to plan our follow-up work here—and maybe someone will have ideas for how to work around it in the interim. Thanks in advance!

**Comment by vladimirtiukhtin** [comment]
@alisdair thanks for the explanation. I believe that's the behavior people would expect. I am happy to share the use-case if you provide a direct link - there are so many topics!

**Comment by vladimirtiukhtin** [comment]
Experimenting further I concluded that lists are not supported at all. Providing a list or a string value to the defaults function results in just different errors

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values. 

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!
